plugin/template/template.go

```diff
@@ -0,0 +1,143 @@
+package template
+
+import (
+	"bytes"
+	"context"
+	"regexp"
+	"strconv"
+	gotmpl "text/template"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// Handler is a plugin handler that takes a query and templates a response.
+type Handler struct {
+	Next      plugin.Handler
+	Templates []template
+}
+
+type template struct {
+	rcode      int
+	class      uint16
+	qtype      uint16
+	regex      []*regexp.Regexp
+	answer     []*gotmpl.Template
+	additional []*gotmpl.Template
+}
+
+type templateData struct {
+	Name     string
+	Regex    string
+	Match    []string
+	Group    map[string]string
+	Class    string
+	Type     string
+	Message  *dns.Msg
+	Question *dns.Question
+}
+
+// ServeDNS implements the plugin.Handler interface.
+func (h Handler) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	for _, template := range h.Templates {
```

Comment:
This needs a check if you should handle the zone in the first place and if not call NextOrFailure here.

Reply:
Where would the plugin get a zone? It handles regex only (for now)

Reply:
```
	zone := plugin.Zones(e.Zones).Matches(state.Name())
	if zone == "" {
		return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
	}
```
as copied from the etcd plugin (there is more code there, take a look)

Reply:
Yes, I saw that, but there is no Zone for this plugin: it acts on regex rather than zone which is a more sophisticated match. There is no `e.Zones` equivalent right now.

If nothing matches `return plugin.NextOrFailure()` is done. If a regex matches the answer will be generated (or a servfail if that causes an error).

Reply:
So the question is mainly: should this plugin actually act on zones and use a zone scope? Or is "try to match everything (the regex can limit the zone anyway)"?

Or should this be changed to act on zones and move the regex into a directive?